src/PowerShellEditorServices/Services/Analysis/PssaCmdletAnalysisEngine.cs

```diff
@@ -347,22 +355,24 @@ private PowerShellResult InvokePowerShell(PSCommand command)
         /// <returns>The output of PowerShell execution.</returns>
         private Collection<PSObject> InvokePowerShellWithModulePathPreservation(System.Management.Automation.PowerShell powershell)
         {
-            if (_alreadyRun)
+            if (_pssaRunCount > PSSA_RUNPOOL_SIZE)
             {
                 return powershell.Invoke();
             }
 
-            // PSScriptAnalyzer uses a runspace pool underneath and we need to stop the PSModulePath from being changed by it
-            try
+            // PSScriptAnalyzer uses a runspace pool underneath and we need to stop the PSModulePath from being changed by it on invocation
+            // Each time a new runspace is opened we need to take this precaution, which is effectively a lock
+            // But when the runspace pool is at capacity and stops opening new runspaces, we can stop
+            using (PSModulePathPreserver.Take())
             {
-                using (PSModulePathPreserver.Take())
+                try
                 {
                     return powershell.Invoke();
                 }
-            }
-            finally
-            {
-                _alreadyRun = true;
+                finally
+                {
+                    Interlocked.Increment(ref _pssaRunCount);
```

Comment:
is it guaranteed that a RunspacePool uses a new Runspace until it hits the max of what it can support?

Reply:
Otherwise the first 10 runs might use the same runspace and then subsequent runs spin up new runspaces.